[MRG + 1] [MAINT] Make sure set_params does not break clone on kernels

MechCoder opened the pull request:
#### Reference Issue

Fixes #6961 
#### What does this implement/fix? Explain your changes.

For example, in kernels such as RBF, using set_params with a
length_scale of different length, used to break clone because of
the non-trivial modification of length_scale in the constructor.

This PR resolves this by moving this logic to **call** and
making hyperparameter attributes properties.

MechCoder committed:
Make sure set_params does not break clone on kernels

For example, in kernels such as RBF, using set_params with a
length_scale of different length, used to break clone because of
the non-trivial modification of length_scale in the constructor.

This PR resolves this by moving this logic to __call__ and
making hyperparameter attributes properties.

MechCoder commented:
cc: @jmetzen @jnothman @amueller

MechCoder commented:
I did not inherit from `BaseEstimator`, since it requires that `fit` etc should be implemented and I assume that was the intention earlier on as well.

jnothman commented:
I wouldn't have thought it sensible to inherit from `BaseEstimator`, although I suppose the `fit` model (i.e. some secondary initialisation phase) would mean properties are unnecessary. I do think the shared `set_params`, `get_params` code should be refactored to `sklearn.base.Parameterised` or something.

MechCoder commented:
I had similar line of thought as well but I prefer not to do it here.

jnothman commented:
That's fine... but why mention `BaseEstimator` then? :p

jnothman commented on sklearn/gaussian_process/tests/test_kernels.py:
```diff
@@ -185,6 +185,7 @@ def test_kernel_stationary():
+    bounds = (1e-5, 1e5)
     for kernel in kernels:
         kernel_cloned = clone(kernel)
 
```
The test below looks at `__dict__` which is now not useful. You'll need to go looking for `__dict__` magic in tests.

MechCoder replied:
Thanks for the catch!

jnothman commented:
Could you please check whether and how similar issues are being tested in `sklearn.tests.test_common`?

MechCoder commented:
Oh, I mentioned it because of the same reason, that it would have been convenient to do away with the `get_params`, `set_params` duplication.

jnothman commented:
Am I right in thinking this is also more *correct* in that the hyperparameter bounds will now be correctly updated with `set_params`?

MechCoder commented:
Sorry, I do not understand. The bounds on the hyperparameters are fixed during instantiation.

MechCoder commented:
I commented too quickly. Yes, you are right.

jnothman commented:
Might as well add a non-regression test for that.

MechCoder commented:
There is one more inconsistent behaviour with sklearn's estimator:

```python
class a(object):
    pass

b = a()
c = a()
b == c
False

el1 = ElasticNet()
el2 = ElasticNet()
el1 == el2
False

rbf1 = RBF()
rbf2 = RBF()
rbf1 == rbf2
True
```

MechCoder committed:
Fix tests by using get_params and implement __eq__ for Hyperparameter

MechCoder commented:
@jnothman I've addressed your comments. I still do not get the corner-case where the bounds on the hyperparameter are not set correctly in master. Would you be able to give an example?

Also, there are no common tests that check the behaviour of `clone` after a non-empty `set_params`

MechCoder committed:
Explain XXX

jnothman commented:
```python
>>> RBF(length_scale_bounds=(1,5)).hyperparameter_length_scale.bounds
array([[1, 5]])
>>> RBF(length_scale_bounds=(1,5)).set_params(length_scale_bounds=(10,50)).hyperparameter_length_scale.bounds
array([[1, 5]])
```

amueller commented:
can you rebase? Is that integration error in master?

amueller commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -92,6 +92,13 @@ def __new__(cls, name, value_type, bounds, n_elements=1, fixed=None):
         return super(Hyperparameter, cls).__new__(
             cls, name, value_type, bounds, n_elements, fixed)
 
+    def __eq__(self, other):
```
Please add a comment why this is needed.

MechCoder replied:
This is not useful internally. I just added it for testing purposes. (So that I can test that the hyperparameters of the parent and cloned kernel are the same directly)

raghavrv replied:
+1 for a comment stating that it is added for testing...

amueller commented:
LGTM

MechCoder committed:
Check that hyperparameters are equal after non-trivial set_params

MechCoder commented:
@jnothman Would be helpful if you could make a final pass.

MechCoder commented:
Not sure, what I broke in the last commit that just changes tests.

MechCoder commented:
OH, seems like my previous travis builds failed. Never mind, it was just me dreaming that all my Travis builds has passed.

MechCoder commented:
I've taken care of the tests.

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -224,23 +234,25 @@ def theta(self, theta):
             raise ValueError("theta has not the correct number of entries."
                              " Should be %d; given are %d"
                              % (i, len(theta)))
+        self.set_params(**params)
```
I don't get why this is preferred over `setattr`, but it's okay...

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1141,14 +1153,23 @@ def __call__(self, X, Y=None, eval_gradient=False):
             is True.
         """
+        length_scale = self.length_scale
+        if np.iterable(length_scale):
```
Is this logic just `np.squeeze(length_scale).astype(np.float)`? Could we factor out `length_scale = _check_length_scale(X, length_scale)`?

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1141,14 +1153,23 @@ def __call__(self, X, Y=None, eval_gradient=False):
         X = np.atleast_2d(X)
-        if self.anisotropic and X.shape[1] != self.length_scale.shape[0]:
+        if self.anisotropic and X.shape[1] != length_scale.shape[0]:
             raise Exception("Anisotropic kernel must have the same number of "
```
Not really your problem, but:

* Why is this `Exception` rather than `ValueError`???
* Elsewhere in the file is "Anisotropic kernels require that the number of length scales and features match". It should match this error message, but I also don't think that code path could ever be executed.

MechCoder replied:
I don't understand:

```python
rbf = RBF(length_scale=[1.0, 2.0])
X = [[1.0, 2.0, 3.0]]
rbf(X)
```

but surely you meant something else?

MechCoder replied:
Oh, so you meant in the other places. Of course..

jnothman replied:
I mean the [else case below](https://github.com/scikit-learn/scikit-learn/pull/6963/files/19d6e2490e25e6670f1b9ab15541a61d0c2052d4#diff-fb7476d7a12141c65a563c589a014cb3R1200) has already been caught out here.

MechCoder replied:
Yes, figured that out, will remove it.

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -1704,18 +1744,15 @@ class PairwiseKernel(Kernel):
 
     def __init__(self, gamma=1.0, gamma_bounds=(1e-5, 1e5), metric="linear",
-                 pairwise_kernels_kwargs=None):
+                 **pairwise_kernels_kwargs):
```
are you sure this is a good decision? If so, the docstring needs to be changed to include `**`; but I also suspect this should be proposed separately.

MechCoder replied:
no, I'm not sure it is a great decision :p

MechCoder replied:
Also, not sure why gamma is treated specially (and which gamma)

jnothman replied:
Not sure either, but clearly gamma is to be a tunable hyperparameter, and all others are not.

jnothman commented:
Otherwise LGTM.

MechCoder committed:
Refactor into _check_length-scale

MechCoder commented:
Thanks for the reviews! Will merge when tests pass.

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -33,6 +33,21 @@
 
 
+def _check_length_scale(X, length_scale):
+    if np.iterable(length_scale):
```
you don't like the `np.squeeze` solution?

jnothman replied:
i.e. `length_scale = np.squeeze(length_scale).astype(float)`

MechCoder replied:
Hmm. I thought about that but it creates a zero-dimensional array for a scalar. Are you sure that will not create any consequences?

jnothman replied:
Not sure, but arithmetic operations involving a 0d-array and a Python scalar produce a Python scalar; those involving a 0d-array and a >0d numpy array produce a numpy array... I doubt there'll be any issue except if there's any type checking (I doubt it), and the tests should tell us if anything obvious breaks.

jnothman replied:
This test just seems extremely verbose...

jnothman replied:
And I don't really get why we're accepting a 1d array of length 1 as a substitute for a scalar anyway...

MechCoder replied:
yeah fine, I just have a phobia for things that look like scalars but are not.

jnothman commented on sklearn/gaussian_process/kernels.py:
```diff
@@ -33,6 +33,21 @@
+            length_scale = float(length_scale[0])
+    else:
+        length_scale = float(length_scale)
+    if np.ndim(length_scale) == 1 and X.shape[1] != length_scale.shape[0]:
```
Is there a policy for `length_scale.ndim > 1`

jnothman commented:
Thanks @mechcoder, this was certainly a tricky one...

On 17 August 2016 at 16:44, Manoj Kumar <notifications@github.com> wrote:

> Merged #6963 <https://github.com/scikit-learn/scikit-learn/pull/6963>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/6963#event-758075145>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz66aEz0zEmRywzB_DljrUuc33GO-Zks5qgq2wgaJpZM4JGyEW>
> .
>
